File: 0139-word-break/0139-word-break.ts
function wordBreak(s: string, wordDict: string[]): boolean {
    const n = s.length
    const set: Set<string> = new Set(wordDict)
    const dp: boolean[] = Array(n+1).fill(false)
    dp[0] = true
    for (let i=1; i <= n; i++) {
        for (let j=0; j < i; j++) {
            const sub = s.substring(j, i)
            if (set.has(sub) && dp[j]) {
                dp[i] = true
            }
        }
    }
    return dp[n]
};

File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        
        n = len(s)
        wordDict = set(wordDict)
        
        dp = [False] * (n + 1)
        dp[0] = True
        
        for i in range(1, n+1):
            for j in range(0, i):
                if s[j:i:] in wordDict and dp[j]:
                    dp[i] = True
                    
        return dp[n]
        